astropy/units/quantity_helper/function_helpers.py

```diff
@@ -1103,3 +1112,120 @@ def unstructured_to_structured(arr, dtype, *args, **kwargs):
+    from astropy.units import StructuredUnit
+
+    # Make Structured unit.
+    units = iterable if isinstance(iterable, StructuredUnit) else StructuredUnit(iterable)
+
+    # Yield from structured unit.
+    for k, v in units.items():
+        if isinstance(v, StructuredUnit):
+            yield from _izip_units_flat(v)
+        else:
+            yield (k, v)
+
+
+def merge_units(*units: UnitBase | StructuredUnit | None) -> StructuredUnit:
+    """Merge a sequence of units into a structured unit.
+
+    .. todo::
+
+        Either make this function public or incorporate into ``StructuredUnit``.
+
+    Parameters
+    ----------
+    *units : `astropy.units.UnitBase` or `astropy.units.StructuredUnit`
+        The units to merge into one structured unit.
+
+    Returns
+    -------
+    `astropy.units.StructuredUnit`
+        Structured unit of constituent unit.
+    """
+    from astropy.units import StructuredUnit, UnitBase
+
+    # filter units, excluding None
+    actual_units = tuple(unit for unit in units if isinstance(unit, (UnitBase, StructuredUnit)))
+    flat = tuple(_izip_units_flat(StructuredUnit(actual_units)))
+    return StructuredUnit(tuple(uu for _, uu in flat), names=tuple(n for n, _ in flat))
+
+
+@function_helper(helps=rfn.merge_arrays)
+def merge_arrays(
+    seqarrays: Quantity | Sequence[Quantity],
+    fill_value: float = -1,
+    flatten: bool = False,
+    usemask: bool = False,
+    asrecarray: bool = False,
+) -> tuple[tuple[Any, ...], dict[str, Any], StructuredUnit, None]:
+    """Merge structured Quantities field by field.
+
+    Parameters
+    ----------
+    seqarrays : sequence of Quantity
+        Sequence of arrays.
+    fill_value : {float}, optional
+        Filling value used to pad missing data on the shorter arrays.
+    flatten : {False, True}, optional
+        Whether to collapse nested fields.
+    usemask : {False, True}, optional
+        Whether to return a masked array or not.
+
+        .. warning::
+            Not yet supported.
+    asrecarray : {False, True}, optional
+        Whether to return a recarray (MaskedRecords) or not.
+
+        .. warning::
+            Not yet supported.
+
+    Returns
+    -------
+    Quantity
+        Merged structured Quantity.
+
+    Raises
+    ------
+    ValueError
+        If ``asrecarray`` is `True`.
+        If ``usemask`` is `True`.
+    """
+    from astropy.units import Quantity, StructuredUnit
+
+    if asrecarray:
+        raise ValueError("asrecarray=True is not supported.")
+    if usemask:
+        raise ValueError("usemask=True is not supported.")
+
+    # Do we have a single ndarray as input ?
+    if isinstance(seqarrays, Quantity):
+        LEN1 = True
+        arrays = (seqarrays.value,)
+        units = (seqarrays.unit,)
+    else:
+        arrays = tuple(a.value for a in seqarrays)
+        units = tuple(q.unit for q in seqarrays)
+        LEN1 = True if len(arrays) == 1 else False
+
+    if flatten:
+        unit = merge_units(*units)
```

Comment:
I think you could do directly `unit  = StructuredUnit(tuple(_izip_units_flat(units)))`. It is not possible for any argument not to have a unit, and the names are not important. Or rather, they are changed when the unit is actually attached to the array:
```
In [23]: a = np.array([(1., 2.), (3., 4.)], dtype=[('a', 'f8'), ('b', 'f8')])

In [24]: unit = u.StructuredUnit(('m', 'cm'))

In [25]: unit.field_names
Out[25]: ('f0', 'f1')

In [26]: q = a << unit

In [27]: q.unit.field_names
Out[27]: ('a', 'b')
```
I think this means `merge_units` can be removed...

Reply:
Great! Although maybe in a followup we could add a classmethod ``from_stuff`` that makes this publicly accessible

```
@classmethod
def from_units(cls, *units: UnitBase | StructuredUnit | None) -> StructuredUnit:
    return cls(tuple(_izip_units_flat(units)))
```

Reply:
But it does work if ``_izip_units_flat`` only returns the value, and not the name.

Reply:
Great that it works. Without a better use case, it is not quite obvious to me that the flattened unit constructor needs to be on the class - though maybe one can just get to concatenate units (with `np.concatenate`?). Anyway, not for this PR!